### PR TITLE
JAMES-3028 Split blobstore cassandra

### PR DESCRIPTION
**Pull request by remk**
replace https://github.com/linagora/james-project/pull/3046

### CODE REVIEW

File: server/blob/blob-api/src/test/java/org/apache/james/blob/api/DumbBlobStoreFixture.java [hunk 1]
```diff
@@ -31,7 +31,7 @@
     byte[] EMPTY_BYTEARRAY = {};
     byte[] SHORT_BYTEARRAY = SHORT_STRING.getBytes(StandardCharsets.UTF_8);
     byte[] ELEVEN_KILOBYTES = Strings.repeat("2103456789\n", 1000).getBytes(StandardCharsets.UTF_8);
-    String TWELVE_MEGABYTES_STRING = Strings.repeat("7893456789\r\n", 1024 * 10);
+    String TWELVE_MEGABYTES_STRING = Strings.repeat("7893456789\r\n", 1024 * 1024); //FIXME
```
Comment (rouazana) [inline server/blob/blob-api/src/test/java/org/apache/james/blob/api/DumbBlobStoreFixture.java]: you can remove the fixme comment then

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStore.java [hunk 2]
```diff
@@ -19,204 +19,131 @@
...
+        Preconditions.checkNotNull(data);
+        HashingInputStream hashingInputStream = new HashingInputStream(Hashing.sha256(), data);
+        return Mono.using(
+            () -> new FileBackedOutputStream(FILE_THRESHOLD),
+            fileBackedOutputStream -> saveAndGenerateBlobId(bucketName, hashingInputStream, fileBackedOutputStream),
+            fileBackedOutputStream -> {
+                try {
+                    fileBackedOutputStream.reset();
+                } catch (IOException e) {
+                    throw new RuntimeException(e);
```
Comment (chibenwa) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStore.java]: Isn't this something we can do with a throwing lambda?

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStore.java [hunk 3]
```diff
@@ -19,204 +19,131 @@
...
+        Preconditions.checkNotNull(bucketName);
+        return retrieveStoredValue(bucketName, blobId);
     }
 
     @Override
     public InputStream read(BucketName bucketName, BlobId blobId) {
-        return ReactorUtils.toInputStream(readBlobParts(bucketName, blobId));
+        Preconditions.checkNotNull(bucketName);
+        return retrieveStoredValue(bucketName, blobId)
+            .map(ByteArrayInputStream::new)
```
Comment (chibenwa) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStore.java]: Are we loosing the `streaming` read capability in the equation?

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/utils/DataChunker.java [hunk 4]
```diff
@@ -19,34 +19,57 @@
...
+        return Flux
+            .<ByteBuffer>generate(sink -> {
+                try {
+                    byte[] buffer = new byte[chunkSize];
+
+                    int size = bufferedInputStream.read(buffer);
+                    if (size <= 0) {
+                        sink.complete();
+                    } else {
+                        sink.next(ByteBuffer.wrap(buffer, 0, size));
```
Comment (chibenwa) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/utils/DataChunker.java]: This should be simplified in `ByteBuffer.wrap(buffer)`

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraDumbBlobStore.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+    private Mono<Integer> saveBlobParts(BucketName bucketName, byte[] data, BlobId blobId) {
+
+        Stream<Pair<Integer, ByteBuffer>> chunks = dataChunker.chunk(data, configuration.getBlobPartSize());
+        return Flux.fromStream(chunks)
+            .publishOn(Schedulers.elastic(), PREFETCH)
+            .flatMap(pair -> writePart(bucketName, blobId, pair.getKey(), pair.getValue())
+                .then(Mono.just(getChunkNum(pair))))
+            .collect(Collectors.maxBy(Comparator.comparingInt(x -> x)))
+            .flatMap(Mono::justOrEmpty)
```
Comment (chibenwa) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraDumbBlobStore.java]: Use handle please!

**Comment by Arsnael** [comment]
CI restart

test this please

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/utils/DataChunker.java (hunk 4 above)
Comment (mbaechler) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/utils/DataChunker.java]: no, you don't  have the guarantee that the size read is what you asked for

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStore.java (hunk 3 above)
Comment (mbaechler) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStore.java]: yes, this method is weird, needs to be fixed

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraDumbBlobStore.java (hunk 5 above)
Comment (remk) [reply server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraDumbBlobStore.java]: this code doesn't exist in later commit. Thanks for the hint thought.

**Comment by remk** [comment]
```
[c48cc88a5579682203c54dd52a03274c19c59fe6] 1) No implementation for org.apache.james.blob.api.BucketName annotated with @com.google.inject.name.Named(value="cassandraDefault") was bound.
[c48cc88a5579682203c54dd52a03274c19c59fe6]   while locating org.apache.james.blob.api.BucketName annotated with @com.google.inject.name.Named(value="cassandraDefault")
[c48cc88a5579682203c54dd52a03274c19c59fe6]     for the 2nd parameter of org.apache.james.blob.cassandra.CassandraBlobStore.<init>(CassandraBlobStore.java:55)
[c48cc88a5579682203c54dd52a03274c19c59fe6]   at org.apache.james.mailbox.cassandra.mail.utils.GuiceUtils.lambda$testInjector$2(GuiceUtils.java:52) (via modules: com.google.inject.util.Modules$CombinedModule -> org.apache.james.mailbox.cassandra.mail.utils.GuiceUtils$$Lambda$547/0x0000000100608440)
[c48cc88a5579682203c54dd52a03274c19c59fe6] 
[c48cc88a5579682203c54dd52a03274c19c59fe6] 2) No implementation for org.apache.james.blob.api.BucketName annotated with @com.google.inject.name.Named(value="cassandraDefault") was bound.
[c48cc88a5579682203c54dd52a03274c19c59fe6]   while locating org.apache.james.blob.api.BucketName annotated with @com.google.inject.name.Named(value="cassandraDefault")
[c48cc88a5579682203c54dd52a03274c19c59fe6]     for the 4th parameter of org.apache.james.blob.cassandra.CassandraDumbBlobStore.<init>(CassandraDumbBlobStore.java:62)
[c48cc88a5579682203c54dd52a03274c19c59fe6]   while locating org.apache.james.blob.cassandra.CassandraDumbBlobStore
[c48cc88a5579682203c54dd52a03274c19c59fe6]     for the 3rd parameter of org.apache.james.blob.cassandra.CassandraBlobStore.<init>(CassandraBlobStore.java:55)
[c48cc88a5579682203c54dd52a03274c19c59fe6]   at org.apache.james.mailbox.cassandra.mail.utils.GuiceUtils.lambda$testInjector$2(GuiceUtils.java:52) (via modules: com.google.inject.util.Modules$CombinedModule -> org.apache.james.mailbox.cassandra.mail.utils.GuiceUtils$$Lambda$547/0x0000000100608440)
[c48cc88a5579682203c54dd52a03274c19c59fe6] 
[c48cc88a5579682203c54dd52a03274c19c59fe6] 2 errors
[c48cc88a5579682203c54dd52a03274c19c59fe6] 	at org.apache.james.mailbox.cassandra.mail.CassandraMessageMapperTest.createMapperProvider(CassandraMessageMapperTest.java:33)
[c48cc88a5579682203c54dd52a03274c19c59fe6] 
```

**Comment by Arsnael** [comment]
```
[634927754e4c85911e76dd4f58e98b3555fdd4cd] [ERROR] Errors: 
[634927754e4c85911e76dd4f58e98b3555fdd4cd] [ERROR]   MemoryDumbBlobStoreTest.mixingSaveReadAndDeleteShouldReturnConsistentState » Execution
```

**Comment by mbaechler** [comment]
> ```
> [634927754e4c85911e76dd4f58e98b3555fdd4cd] [ERROR] Errors: 
> [634927754e4c85911e76dd4f58e98b3555fdd4cd] [ERROR]   MemoryDumbBlobStoreTest.mixingSaveReadAndDeleteShouldReturnConsistentState » Execution
> ```

More precisely 
```
[634927754e4c85911e76dd4f58e98b3555fdd4cd] java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: Java heap space
[634927754e4c85911e76dd4f58e98b3555fdd4cd] Caused by: java.lang.OutOfMemoryError: Java heap space
```

**Comment by remk** [comment]
```
[634927754e4c85911e76dd4f58e98b3555fdd4cd] [INFO] -------------------------------------------------------
[634927754e4c85911e76dd4f58e98b3555fdd4cd] [INFO] Running org.apache.james.blob.memory.MemoryDumbBlobStoreTest
[634927754e4c85911e76dd4f58e98b3555fdd4cd] [ERROR] Tests run: 64, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 14.595 s <<< FAILURE! - in org.apache.james.blob.memory.MemoryDumbBlobStoreTest
[634927754e4c85911e76dd4f58e98b3555fdd4cd] [ERROR] mixingSaveReadAndDeleteShouldReturnConsistentState  Time elapsed: 6.624 s  <<< ERROR!
[634927754e4c85911e76dd4f58e98b3555fdd4cd] java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: Java heap space
[634927754e4c85911e76dd4f58e98b3555fdd4cd] Caused by: java.lang.OutOfMemoryError: Java heap space
[634927754e4c85911e76dd4f58e98b3555fdd4cd] 
[634927754e4c85911e76dd4f58e98b3555fdd4cd] [INFO] Running org.apache.james.blob.memory.MemoryBlobStoreTest
[634927754e4c85911e76dd4f58e98b3555fdd4cd] [INFO] Tests run: 77, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.769 s - in org.apache.james.blob.memory.MemoryBlobStoreTest
[634927754e4c85911e76dd4f58e98b3555fdd4cd] [INFO] 
[634927754e4c85911e76dd4f58e98b3555fdd4cd] [INFO] Results:
[634927754e4c85911e76dd4f58e98b3555fdd4cd] [INFO] 
[634927754e4c85911e76dd4f58e98b3555fdd4cd] [ERROR] Errors: 
[634927754e4c85911e76dd4f58e98b3555fdd4cd] [ERROR]   MemoryDumbBlobStoreTest.mixingSaveReadAndDeleteShouldReturnConsistentState » Execution
```

**Comment by mbaechler** [comment]
Replaced by https://github.com/linagora/james-project/pull/3087

**Pull request closed without merging**